README.md
# Salvage Quickstart

You got to this planet somehow. That implies a highly advanced technological civilization. Why shouldn't you start with a bit of that advanced technology?

In this mod, you start the game with a limited number of extremely powerful devices salvaged from your crashed spaceship. All have an internal fusion power supply that can supply its own needs without electricity or fuel provided from the outside.

I intended each of these devices to do the following:

1. Reduce the tedium of the early game

1. Still incentivize the player to build additional infrastructure as quickly as possible.

1. Give a taste of the toys that await at the bottom of the tech tree.

1. Never become useless.

Practically speaking, the roboport and construction robots are
significantly better than the construction robots you can eventually
research. However, they provide and use much less power, making the
starter roboport less effective than normal roboports once you have
access to them. The starter robots can use normal roboports, but you
start with a limited number and can never get more. Losses to biters or
carelessly thrown grenades will deny you their use forever.

The mining drill and furnace are very fast, but have a limited inventory
and do not output to belts, initially requiring players to manually remove
and feed ore to maintain throughput. They can be automated with inserters,
but there is very limited space for them, severely restricting throughput,
although they get better as inserters improve.


changelog.txt
---------------------------------------------------------------------------------------------------
Version: 0.4.0
Date: 2019-08-19
  Compatibility:
    - Update to Factorio 0.18.
  Balancing:
    - Significantly increase the range of the starting roboport. Now it is truly something you can set up once and never move.
    - Increase starting construction robot speed.
    - Increase size of the starting miner inventory to make it slightly less annoying, but still not a replacement for real mining infrastructure.
---------------------------------------------------------------------------------------------------
Version: 0.3.3
Date: 2019-08-19
  Bugfixes:
    - Fixed crash at start of multiplayer game.
---------------------------------------------------------------------------------------------------
Version: 0.3.2
Date: 2019-07-29
  Bugfixes:
    - Fixed wreckage resistances.
  Changes:
    - It is now possible to select the mining drill's edges to see its statistics and the remaining resources underneath.
---------------------------------------------------------------------------------------------------
Version: 0.3.1
Date: 2019-07-27
  Bugfixes:
    - Fix typo that caused a crash when mining the roboport.
    - Removed some stray debug logging.
    - Fixed that wreckage contained more construction robots than intended.
---------------------------------------------------------------------------------------------------
Version: 0.3.0
Date: 2019-07-27
  Changes:
    - Wreckage can now be mined for materials.
    - Wreckage containing machines now has resistance to fire and physical so it isn't accidentally destroyed.
    - Added note in startup message that explosives can be used to remove big wreckage.
    - Players now start with wood in inventory as in vanilla, for those zero tree challenge runs.
---------------------------------------------------------------------------------------------------
Version: 0.2.0
Date: 2019-07-26
  Changes:
    - Added a crash sequence with falling flaming debris. Advanced machines are contained in the wreckage.
    - Moved mining drill chest from northwest corner to southeast corner.
    - The entire mining drill can be clicked to access the chest.
    - Renamed furnace to make its purpose clearer.
---------------------------------------------------------------------------------------------------
Version: 0.1.0
Date: 2019-07-24
  Features:
    - Self-powered 2x2 furnace, 10x as fast as a steel/electric furnace. Not craftable.
    - Self-powered 2x2 mining drill that outputs only to a built-in chest. It is possible to pull from the chest with 2 inserters. Not craftable.
    - The player starts with 10 very fast construction robots that consume little energy. Not craftable.
    - Self-powered 2x2 roboport with built-in logistic storage chest. Slow charging rate and 2 charging slots, which should be plenty for the 10 starter construction bots. Not craftable.


control.lua
local crashfx = require "crashfx"

local function moveposition(position, offset)
  return {x=position.x + offset.x, y=position.y + offset.y}
end

local function offset(direction, longitudinal, orthogonal)
  if direction == defines.direction.north then
    return {x=orthogonal, y=-longitudinal}
  end

  if direction == defines.direction.south then
    return {x=-orthogonal, y=longitudinal}
  end

  if direction == defines.direction.east then
    return {x=longitudinal, y=orthogonal}
  end

  if direction == defines.direction.west then
    return {x=-longitudinal, y=-orthogonal}
  end
end

local function on_built_entity(event)
  local entity = event.created_entity or event.entity
  local chest_name
  local off
  if entity.name == "sqs-mining-drill" then
    entity.direction = defines.direction.south
    local chest = entity.surface.create_entity{
      name = "sqs-mining-drill-chest",
      force = entity.force,
      position = moveposition(entity.position, { x = 0.5, y = 0.5 }),
    }
  elseif entity.name == "sqs-roboport" then
    local chest = entity.surface.create_entity{
      name = "logistic-chest-storage",
      force = entity.force,
      position = moveposition(entity.position, { x = 0.5, y = 0.5 }),
    }
    chest.destructible = false
    chest.minable = false
  end
end

local function on_mined_entity(event)
  local entity = event.entity
  if entity.name == "sqs-mining-drill-chest" then
    local drill = entity.surface.find_entity(
      "sqs-mining-drill",
      moveposition(entity.position, {x = -0.5, y = -0.5}))
    if drill then drill.destroy() end
  elseif entity.name == "sqs-roboport" then
    local chest = entity.surface.find_entity(
      "logistic-chest-storage",
      moveposition(entity.position, {x = 0.5, y = 0.5}))
    local buffer = event.buffer
    if chest then
      if buffer then
        local inventory = chest.get_inventory(defines.inventory.chest)
        for i=1,#inventory do
          local stack = inventory[i]
          if stack.valid_for_read then
            buffer.insert(stack)
            stack.clear()
          end
        end
      end
      chest.destroy()
    end
  end
end

local CRASH_ITEMS = {
  {"sqs-mining-drill"},
  {"sqs-furnace"},
  {"sqs-roboport", "sqs-construction-robot"},
}

local function on_tick(event)
  if event.tick == 0 then
    if game.is_multiplayer() then
      game.print({"sqs-intro-msg"})
    else
      game.show_message_dialog{text = {"sqs-intro-msg"}}
    end
  end

  local container = crashfx.run(event.tick)
  if container then
    local item_names = CRASH_ITEMS[global.crashfx.containers_spawned]
    for _, item_name in pairs(item_names) do
      container.insert(item_name)
    end
  elseif container == false then
    --log("crash sequence ended on tick "..event.tick)
    script.on_event(defines.events.on_tick, nil)
  end
end

local handlers = {
  on_built_entity = on_built_entity,
  on_player_mined_entity = on_mined_entity,
  on_robot_built_entity = on_built_entity,
  on_robot_mined_entity = on_mined_entity,
  script_raised_destroy = on_mined_entity,
  script_raised_revive = on_built_entity,
}

for name, handler in pairs(handlers) do
  script.on_event(defines.events[name], handler)
end

local function on_load()
  if not (global.crashfx and global.crashfx.done) then
    script.on_event(defines.events.on_tick, on_tick)
  end
end

local function on_init()
  if remote.interfaces["freeplay"] then
    remote.call("freeplay", "set_skip_intro", true)
    local starting = remote.call("freeplay", "get_created_items")
    starting["burner-mining-drill"] = nil
    starting["stone-furnace"] = nil
    remote.call("freeplay", "set_created_items", starting)
  end
  on_load()
end

script.on_init(on_init)
script.on_load(on_load)

script.on_configuration_changed(function(data)
  if data.mod_changes
  and data.mod_changes[script.mod_name]
  and data.mod_changes[script.mod_name].old_version == "0.1.0" then
    global.crashfx = { containers_spawned = 3 }
    for _, s in pairs(game.surfaces) do
      for _, chest in pairs(s.find_entities_filtered{name = "sqs-mining-drill-chest"}) do
        chest.minable = true
        chest.teleport(moveposition(chest.position, { x = 1, y = 1 }))
      end
    end
  end
end)



crashfx.lua
local END_TIME = 30 * 60
local DECAY = 300
local CENTER_STDDEV = 15
local NUM_CONTAINERS = 3
local CONTAINER_PROB = 0.01
local CONTAINER_STDDEV = 10
local FIRE_PROB = 0.7
local SAFE_RANGE = 4
local WRECKAGE_FORCE = "neutral"

local WRECKS = {
  { prob = 0.1, name = "sqs-medium-ship-wreck", explosion = "ground-explosion", stddev = 20 },
  { prob = 0.5, name = "sqs-small-ship-wreck" , explosion = "big-explosion"   , stddev = 50 },
  { prob = 3 , name = "sqs-small-scorchmark" , explosion = "explosion"       , stddev = 200 },
}

local ln = math.log
local random = math.random

local center
local nauvis

-- returns a point sampled from a normal distribution centered at 0,0
-- with the provided standard deviation, by the https://en.wikipedia.org/wiki/Marsaglia_polar_method
local function gaussian_point(stddev)
  local x, y = random() * 2 - 1, random() * 2 - 1
  local s = x * x + y * y
  if s >= 1 then return gaussian_point(stddev) end
  local scale = (-2 * ln(s) / s) ^ 0.5 * stddev
  return { x = x * scale, y = y * scale }
end

local function random_position(stddev)
  center = global.crashfx.center
  if not center then
    center = gaussian_point(CENTER_STDDEV)
    global.crashfx.center = center
  end

  local offset = gaussian_point(stddev)
  return { x = center.x + offset.x, y = center.y + offset.y }
end

local AVOID_CHARACTERS = {"character", "container"}
local AVOID_RESOURCES = {"character", "container", "resource"}
local avoid_types_cache = {}
local function avoid_types(name)
  local to_avoid = avoid_types_cache[name]
  if not to_avoid then
    local proto = game.entity_prototypes[name]
    if proto.type == "container" then
      to_avoid = AVOID_RESOURCES
    else
      to_avoid = AVOID_CHARACTERS
    end
    avoid_types_cache[name] = to_avoid
  end
  return to_avoid
end

local function safe_position(name, stddev)
  local pos = random_position(stddev)
  if nauvis.can_place_entity{ name = "fish", position = pos} then
    return safe_position(name, stddev)
  end
  local colliding = nauvis.find_entities_filtered{
    type = avoid_types(name),
    area = {{pos.x-SAFE_RANGE,pos.y-SAFE_RANGE},{pos.x+SAFE_RANGE,pos.y+SAFE_RANGE}},
    limit = 1,
  }
  if colliding[1] then
    --[[
    log(serpent.line
    {
      msg="retry",
      name=name,
      cause_name = colliding[1].name,
      cause_position = colliding[1].position,
    })
    ]]
    return safe_position(name, stddev) end
  return pos
end

local function spawn_container()
  if not global.crashfx then global.crashfx = {} end
  global.crashfx.containers_spawned = (global.crashfx.containers_spawned or 0) + 1

  local name = "big-ship-wreck-"..global.crashfx.containers_spawned
  local position = safe_position(name, CONTAINER_STDDEV)
  local container = nauvis.create_entity{
    name = name,
    position = position,
    force = "neutral",
  }
  nauvis.create_entity{
    name = "massive-explosion",
    position = position,
    force = "neutral",
  }
  nauvis.create_entity{
    name = "ground-explosion",
    position = position,
    force = "neutral",
  }

  return container
end

local function spawn_wreckage(wreck)
  local position = safe_position(wreck.name, wreck.stddev)
  nauvis.create_entity{
    name = wreck.name,
    force = "neutral",
    position = position,
  }
  nauvis.create_entity{
    name = wreck.explosion,
    force = "neutral",
    position = position,
  }
  if random() < FIRE_PROB then
    nauvis.create_entity{
      name = "fire-flame-on-tree",
      position = position,
    }
  end
end

local function spawn_small()
end

local function run(tick)
  if not global.crashfx then
    global.crashfx = { containers_spawned = 0 }
  end
  if global.crashfx.containers_spawned >= NUM_CONTAINERS and tick > END_TIME then
    global.crashfx.done = true
    return false
  end
  if not nauvis then nauvis = game.surfaces.nauvis end

  local r = math.random()
  local time_factor = math.exp(-tick / DECAY)
  if r < CONTAINER_PROB and global.crashfx.containers_spawned < NUM_CONTAINERS then return spawn_container() end
  for i=1,#WRECKS do
    local wreck = WRECKS[i]
    if r < wreck.prob * time_factor then return spawn_wreckage(wreck) end
  end
end

return {
  run = run,
}


prototypes/items.lua
data:extend
{
  {
    type = "item",
    name = "sqs-mining-drill",
    icon = "__salvage-quickstart__/graphics/icons/sqs-mining-drill.png",
    icon_size = 32,
    subgroup = "extraction-machine",
    order = "sqs-mining-drill",
    place_result = "sqs-mining-drill",
    stack_size = 1,
  },
  {
    type = "item",
    name = "sqs-furnace",
    icon = "__salvage-quickstart__/graphics/icons/sqs-furnace.png",
    icon_size = 32,
    subgroup = "smelting-machine",
    order = "sqs-furnace",
    place_result = "sqs-furnace",
    stack_size = 1,
  },
  {
    type = "item",
    name = "sqs-roboport",
    icon = "__salvage-quickstart__/graphics/icons/sqs-roboport.png",
    icon_size = 32,
    subgroup = "logistic-network",
    order = "sqs-roboport",
    place_result = "sqs-roboport",
    stack_size = 1,
  },
  {
    type = "item",
    name = "sqs-construction-robot",
    icon = "__salvage-quickstart__/graphics/icons/sqs-construction-robot.png",
    icon_size = 32,
    subgroup = "logistic-network",
    order = "sqs-roboport",
    place_result = "sqs-construction-robot",
    stack_size = 10,
  },
}


prototypes/wreckage.lua
require "util"

local wreckages = {
  { name = "medium-ship-wreck", minable = { mining_time = 2, results = {
    { type = "item", name = "copper-cable", probability = .80, amount_min = 1, amount_max = 10 },
    { type = "item", name = "steel-plate", probability = .60, amount_min = 1, amount_max = 4 },
  }}},
  { name = "small-ship-wreck", minable = { mining_time = 1, results = {
    { type = "item", name = "steel-plate", probability = .20, amount_min = 1, amount_max = 2 },
  }}},
  { name = "small-scorchmark", results = nil },
}

local function add_resistance(proto, type, decrease, percent)
  if not proto.resistances then proto.resistances = {} end
  for _, resist in pairs(proto.resistances) do
    if resist.type == type then
      resist.decrease = (resist.decrease or 0) < decrease and decrease or resist.decrease
      resist.percent = (resist.percent or 0) < percent and percent or resist.percent
      return
    end
  end
  table.insert(proto.resistances, { type = type, decrease = decrease, percent = percent })
end

local function add_resistances(proto)
  add_resistance(proto, "physical", 10, 90)
  add_resistance(proto, "fire", 0, 100)
  proto.max_health = (proto.max_health < 250) and 250 or proto.max_health
end

local types = { "container", "simple-entity", "corpse" }
local function get_wreck(name)
  for _, type in pairs(types) do
    local proto = data.raw[type][name]
    if proto then return proto end
  end
end

for _, wreck_info in pairs(wreckages) do
  local name = wreck_info.name
  local sqs_wreck = util.merge{get_wreck(name), wreck_info}
  sqs_wreck.name = "sqs-"..sqs_wreck.name
  data:extend{sqs_wreck}
end

add_resistances(data.raw.container["big-ship-wreck-1"])
add_resistances(data.raw.container["big-ship-wreck-2"])
add_resistances(data.raw.container["big-ship-wreck-3"])


prototypes/entities.lua
local function foreach_sprite(t, f)
  if t.filename and t.filename:find("%.png$") then
    f(t)
  end
  for k, v in pairs(t) do
    if type(v) == "table" then
      foreach_sprite(v, f)
    end
  end
end

local function add_tint(t, tint)
  foreach_sprite(t, function(s)
    s.tint = tint
  end)
end

local function change_scale(t, factor)
  foreach_sprite(t, function(s)
    s.scale = (s.scale or 1) * factor
    if s.shift then
      s.shift = util.mul_shift(s.shift, factor)
    end
  end)
end

local drill_chest = table.deepcopy(data.raw.container["steel-chest"])
drill_chest.name = "sqs-mining-drill-chest"
drill_chest.order = "sqs-mining-drill-chest"
drill_chest.minable.result = "sqs-mining-drill"
drill_chest.placeable_by = {item = "sqs-mining-drill", count = 1}
drill_chest.collision_box = {{-0.3, -0.3}, {0.3, 0.3}}
drill_chest.selection_box = {{-1.3, -1.3}, {0.5, 0.5}}
drill_chest.selection_priority = 60
drill_chest.fast_replaceable_group = nil
drill_chest.inventory_size = 4

local drill = table.deepcopy(data.raw["mining-drill"]["burner-mining-drill"])
drill.name = "sqs-mining-drill"
drill.icon = "__salvage-quickstart__/graphics/icons/sqs-mining-drill.png"
table.insert(drill.flags, "not-rotatable")
drill.minable = nil
drill.energy_source = { type = "void" }
drill.mining_speed = 10
drill.vector_to_place_result = {0.5, 0.5}
drill.working_sound.sound.filename = "__base__/sound/fast-transport-belt.ogg"
drill.animations.north = drill.animations.south
add_tint(drill, {r=1, g=0.8, b=0.8})

local furnace = table.deepcopy(data.raw["furnace"]["electric-furnace"])
furnace.name = "sqs-furnace"
furnace.icon = "__salvage-quickstart__/graphics/icons/sqs-furnace.png"
furnace.type = "assembling-machine"
furnace.minable.result = "sqs-furnace"
furnace.corpse = "medium-small-remnants"
furnace.energy_source = { type = "void" }
furnace.collision_box = {{-0.8, -0.8}, {0.8, 0.8}}
furnace.selection_box = {{-1, -1}, {1, 1}}
furnace.crafting_speed = 20
for _, vis in pairs(furnace.working_visualisations) do
  vis.animation.animation_speed = 0.06
  vis.animation.hr_version.animation_speed = 0.06
end
furnace.working_sound.sound.filename = "__base__/sound/accumulator-idle.ogg"
change_scale(furnace, 2/3)
add_tint(furnace, {r=1, g=0.8, b=0.8})

local roboport = table.deepcopy(data.raw["roboport"]["roboport"])
roboport.name = "sqs-roboport"
roboport.icon = "__salvage-quickstart__/graphics/icons/sqs-roboport.png"
roboport.minable.result = "sqs-roboport"
roboport.corpse = "medium-small-remnants"
roboport.energy_source = { type = "void" }
roboport.recharge_minimum = "0J"
roboport.collision_box = {{-0.8, -0.8}, {0.8, 0.8}}
roboport.selection_box = {{-1, -1}, {1, 1}}
roboport.energy_usage = "500kW"
roboport.charging_energy = "250kW"
roboport.construction_radius = 1024
roboport.logistics_radius = 1
roboport.charge_approach_distance = 3
roboport.robot_slots_count = 1
roboport.material_slots_count = 1
roboport.charging_offsets = {{-1, 0}, {1, 0}}
change_scale(roboport, 0.5)
add_tint(roboport, {r=0.8, g=0.8, b=1})

local bot = table.deepcopy(data.raw["construction-robot"]["construction-robot"])
bot.name = "sqs-construction-robot"
bot.icon = "__salvage-quickstart__/graphics/icons/sqs-construction-robot.png"
bot.minable.result = "sqs-construction-robot"
bot.max_health = 500
bot.max_payload_size = 3
bot.speed = 0.30
bot.max_energy = "300kJ"
bot.energy_per_tick = "10J"
bot.energy_per_move = "500J"
change_scale(bot, 0.75)
add_tint(bot, {r=0.8, g=0.8, b=1})

data:extend{drill_chest, drill, furnace, roboport, bot}
